### PR TITLE
Adding tsoa

### PR DESCRIPTION
**Pull request by ana2201**
- POST - create user
- remove `routes` file, generate using tsoa
- create services: `user-service`, `message-service`
- adapt controllers

### CODE REVIEW

File: src/services/users-service.ts [hunk 1]
```diff
@@ -0,0 +1,30 @@
+import { PrismaClient } from '@prisma/client';
+
+import { User } from '../controllers/user';
+
+const prisma = new PrismaClient();
+
+export type UserCreationParams = Pick<User, 'email' | 'name' | 'lastname' | 'password'>;
+
+export class UsersService {
+  public async get(id: number): Promise<User | null> {
```
Comment (aperezcastagnino) [inline src/services/users-service.ts]: we can use this method as static

File: src/controllers/users-controller.ts [hunk 2]
```diff
@@ -1,20 +1,29 @@
...
-    const u = await prisma.user.findUnique({
-      where: {id}
-    });
-
-    res.send(u ? `User information --> ${JSON.stringify(u)}` : 'the user does not exist');
-  } catch (error) {
-    res.status(400).send('The id must be a number'); 
+@Route('users')
+export class UsersController extends Controller {
+  @Get('{userID}')
```
Comment (aperezcastagnino) [inline src/controllers/users-controller.ts]: userId instead of userID

File: src/controllers/users-controller.ts [hunk 3]
```diff
@@ -1,20 +1,29 @@
...
+  public async getUser(
+    @Path() userID: number
+  ): Promise<User | null> {
+    return new UsersService().get(userID);
```
Comment (aperezcastagnino) [inline src/controllers/users-controller.ts]: getUser instead of get

File: src/middleware/error-handler.ts [hunk 4]
```diff
@@ -0,0 +1,34 @@
...
+import { ApiError } from '../utils/apiError';
+import { ErrorInterface } from '../types/error';
+import { errors } from '../utils/errors';
+
+export function notFoundHandler(_req: Request,res: Response) {
+  res.status(404).send({  message: 'Page not found' });
+}
+
+export function errorHandler(
+  err: unknown,
```
Comment (guillezorron) [inline src/middleware/error-handler.ts]: what's this?

File: src/services/users-service.ts [hunk 5]
```diff
@@ -0,0 +1,40 @@
...
+
+export class UsersService {
+  static async getUser(id: number): Promise<User> {
+    const user = await prisma.user.findUnique({ where: { id } });
+      
+    if (!user) {
+      throw new ApiError(errors.NOT_FOUND_USER);
+    }
+
+    return user;
```
Comment (guillezorron) [inline src/services/users-service.ts:17]: What about the status of the request here?

File: src/services/users-service.ts [hunk 6]
```diff
@@ -0,0 +1,40 @@
...
+    try {
+      const user = await prisma.user.create({
+        data: {
+          name: userCreationParams.name,
+          lastname: userCreationParams.lastname,
+          email: userCreationParams.email,
+          password: userCreationParams.password
+        }
+      });
+
```
Comment (guillezorron) [inline src/services/users-service.ts:30]: Same as above

File: src/services/users-service.ts [hunk 7]
```diff
@@ -0,0 +1,40 @@
...
+          lastname: userCreationParams.lastname,
+          email: userCreationParams.email,
+          password: userCreationParams.password
+        }
+      });
+  
+      return user;
+    } catch (e) {
+      if (e instanceof Prisma.PrismaClientKnownRequestError && e.code === 'P2002') { // Prisma error: Unique constraint failed on the {constraint}
+        throw new ApiError(errors.USER_CREATION_CONFLICT);
```
Comment (guillezorron) [inline src/services/users-service.ts:34]: I prefer to use 400 or 422 here. I think is more accurated

File: src/utils/errors.ts [hunk 8]
```diff
@@ -0,0 +1,22 @@
...
+  NOT_FOUND_USER: {
+    httpCode: 404,
+    errorCode: 404_000,
+    description: 'User not found',
+  },
+  USER_CREATION_CONFLICT: {
+    httpCode: 409,
+    errorCode: 409_000,
+    description: 'User already exists',
+  },
```
Comment (guillezorron) [inline src/utils/errors.ts]: I think this error doesn't make sense

**Review by aperezcastagnino — APPROVED** [review]

File: src/middleware/error-handler.ts (hunk 4 above)
Comment (ana2201) [reply src/middleware/error-handler.ts]: This should be type `any`

File: src/services/users-service.ts (hunk 5 above)
Comment (ana2201) [reply src/services/users-service.ts:17]: If there's no error, then the response's status is 200 OK

File: src/services/users-service.ts (hunk 6 above)
Comment (ana2201) [reply src/services/users-service.ts:30]: Same here, if f there's no error on the service level, then the response's status is 200 OK

**Pull request merged**